CV2-5731 Refactoring smooch search

melsawy opened the pull request:
## Description

Refactoring smooch search to return explainers in addition to fact-checks and also to query from a ProjectMedia.

Should implement two methods for this refacto

- [ ] Method 1: method that calls the existing search_for_similar_published_fact_checks_no_cache method but also returns explainers if no fact-check is found.
- [ ] Method 2: An instance method in ProjectMedia model to call Method 1.

References: CV2-5731

## How has this been tested?

Implemented automated tests.

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [ ] I have added regression tests, if the PR fixes a bug
- [ ] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [ ] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [ ] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

melsawy commented on app/models/project_media.rb:
```diff
@@ -553,6 +553,12 @@ def add_nested_objects(ms)
     ms.attributes[:requests] = requests
   end
 
+  def self.get_similar_articles(project_media)
```
Done

caiosba commented on app/models/project_media.rb:
```diff
@@ -553,6 +553,23 @@ def add_nested_objects(ms)
+    # Transcription for UploadedVideo , UploadedAudio and UploadedImage
+    # Title and/or description for Link
+    media = self.media
+    search_query = case media.types
```
This should be `type`, not `types`. If this was not raised by a unit test, please add one.

melsawy replied:
Did not add a unit test for this one but also I see no tests run for the PR is this normal?

caiosba replied:
I think it's because the target branch is not `develop`? But it should run. @chinelo-obitube, can you please check?

caiosba commented on app/models/team.rb:
```diff
@@ -560,6 +560,20 @@ def filter_by_keywords(query, filters, type = 'FactCheck')
+    unless result_ids.blank?
+      # I depend on FactCheck to filter result instead of report_design
+      items = FactCheck.where(report_status: 'published').joins(:claim_description)
+      .joins("INNER JOIN project_medias pm ON pm.id = claim_descriptions.project_media_id ")
```
When possible, it's more clear to use `ActiveRecord` syntax. So, I suggest:

`joins(claim_description: :project_media)`.

Instead of:

`.joins("INNER JOIN project_medias pm ON pm.id = claim_descriptions.project_media_id ")`.

melsawy replied:
Done

caiosba reviewed (approved):
Thanks for addressing the comments from the previous review, @melsawy - please make sure we're keeping the 100% coverage before merging.

melsawy commented:
> Thanks for addressing the comments from the previous review, @melsawy - please make sure we're keeping the 100% coverage before merging.

Sure Caio
